blocks/extensions/plot.py

```diff
@@ -136,14 +130,103 @@ def preexec_fn():
+        display confidence intervals representing the standard deviation of
+        the channel over the dataset.
+
+        .. warning::
+            This extension assumes that a bokeh server is running at the
+            provided url and won't try to start a server autonomously. To
+            start a bokeh server: `bokeh-server --ip=<ip> --port=<port>`.
+            If you want the server to accept connections from any address
+            use <ip>= 0.0.0.0.
+
+        Parameters
+        ----------
+        document : string
+            The name of the document.
+        channels : list
+            The channels to be displayed as a histogram.
+        server_url : string
+            The url of the server.
+
+        """
+        self.document = document
+        self.channels = channels
+        self.server_url = server_url
+
+        kwargs.setdefault('after_epoch', True)
+        kwargs.setdefault("before_first_epoch", False)
+        super(PlotHistogram, self).__init__(**kwargs)
+
+    def do(self, which_callback, *args):
+        channels = self.channels
+        log = self.main_loop.log
+        epoch = log.status['epochs_done']
+
+        # create a new document --> a new session
+        self.session = Session(
+            root_url=self.server_url,
+            load_from_config=False)
+
+        output_server(self.document+' (epoch'+str(epoch)+')',
+                      session=self.session, clear=True)
+
+        for key, value in log.current_row.items():
+            if key in channels:
+                kl_mean = value.mean(axis=0)
+                kl_std_dev = value.std(axis=0)
+                nhid = len(kl_mean)
+
+                fig = figure(title=key, plot_width=1000)
+                # display alternating color histograms
+                fig.quad(top=kl_mean[::2], bottom=0, left=range(nhid)[::2],
+                         right=range(nhid+1)[1::2], fill_color='#66E066',
+                         line_color="black")
+                fig.quad(top=kl_mean[1::2], bottom=0, left=range(nhid)[1::2],
+                         right=range(nhid+1)[2::2], fill_color='#66C2FF',
+                         line_color="black")
+
+                # display confidence intervals (std_dev)
+                # intervals' max height to 20% of the figure
+                kl_std_dev = kl_std_dev.astype(float) / max(
+                    kl_std_dev).astype(float)
+                kl_std_dev = kl_std_dev * max(kl_mean)/5.0
+                # position relatively to the previous histogram
+                std_hist_bottom = kl_mean - kl_std_dev/2.0
+                std_hist_top = kl_mean + kl_std_dev/2.0
+                fig.quad(top=std_hist_top, bottom=std_hist_bottom,
+                         left=[el + 0.47 for el in range(nhid)],
+                         right=[el + 0.53 for el in range(nhid)],
+                         fill_color='black', line_color="black")
+        push()
+
     def __setstate__(self, state):
```

Comment:
Seems like me that you can get rid of `__getstate__` and `__setstate__` entirely here. `sub` used to contain the subprocess, but you don't have it here, so the default behavior of `pickle` should work.